### DISCUSSION TITLE
[ListView] Events detached from 'SelectionChanged' are still invoked

### DISCUSSION THREAD

**Discussion by marwalsch** [unoplatform/uno#8525] [Q&A] [answered]
### Current behavior

When an event is detached from a `ListView`'s `SelectionChanged`, it is still invoked. This behaviour is not reproducible on most regular controls, but it might be for other events on the `ListView` (haven't tested it yet).  I suspect this is related to Uno relying on the native `ListView`.

```
xListView.SelectionChanged -= OnSelectionChanged;

xListView.SelectedIndex = 1;

xListView.SelectionChanged += OnSelectionChanged;
```

When the above snippet is run, `OnSelectionChanged` will still be executed, provided it has been attached at some point earlier.

### Expected behavior

The event should not be invoked.
or
An warning should indicate that events can't be detached on that particular control.

### How to reproduce it (as minimally and precisely as possible)

1. Run the sample.
2. Select any item and witness the event firing.
3. Press the Button; 
   - provided index 1 was not selected already, it will be changed to programmatically;
   - during this process the event is temporarily detached;
   - on Windows: you won't see the event firing;
   - on iOS/Android you will see the event firing.

### Workaround

This approach seems to be one of [the few viable approaches ](https://stackoverflow.com/questions/7412280/determine-whether-selector-selectionchanged-event-was-initiated-by-a-user)to check if `SelectionChanged` was invoked by user input. In that case one might rely to wrapping `ListItem`s in `Button`s and listen to the `Click` event.

### Works on UWP/WinUI

Yes

### Environment

Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia

### NuGet package version(s)

- **Uno.WinUI** 4.0.11

### Affected platforms

iOS, Android

### IDE

Visual Studio 2022

### IDE version

17.1.3

### Relevant plugins

None

### Anything else we need to know?

[8541-Repro.zip](https://github.com/unoplatform/uno/files/8443543/8541-Repro.zip)

**Comment by jeromelaban** [comment, marked as answer]
Thanks for the report.

This behavior is actually related to the way events are handled in XAML. You'll notice that when looking at the stack trace, there's an additional generated method which is present:

```csharp
var SelectionChanged_OnSelectionChanged_That = 
   (this as global::Uno.UI.DataBinding.IWeakReferenceProvider).WeakReference;
c4.SelectionChanged += (OnSelectionChanged_sender,OnSelectionChanged_e) =>
   (SelectionChanged_OnSelectionChanged_That.Target as MainPage)?.OnSelectionChanged(OnSelectionChanged_sender,OnSelectionChanged_e);
```

Which is why when you execute:
```csharp
xListView.SelectionChanged -= OnSelectionChanged;
```
the provided event handler is not removed because it is never registered directly, as instead it's invoked through a generated lambda.

This way of handling events is needed for memory considerations.

If you want to be able to unregister the event, you'll need to explicitly register the handler in the `Page` constructor, and remove the XAML based registration.

Now, if you want to avoid reentrancy, you may handle change a state flag from the entry points (`OnReassign` in your sample) that cause the changes and ignore some processing in the `OnSelectionChanged` code. It may become trickier if you use bindings though, but that's probably another question :)

**Reply by marwalsch** [reply]
Thanks a lot for the explanation,

calling `xListView.SelectionChanged += OnSelectionChanged;` did it!

> Now, if you want to avoid reentrancy, you may handle change a state flag from the entry points (OnReassign in your sample) that cause the changes and ignore some processing in the OnSelectionChanged code

I agree on the possibility this approach for the repro,  though temporarily suspending the event can come in handy to make painfully convoluted solutions a bit smoother.

One thing boggles my mind though:

According to the [docs](https://docs.microsoft.com/en-us/windows/uwp/xaml-platform/events-and-routed-events-overview) the ability to unsubscribe XAML-registered events is API dependent:

> You can also remove handlers for cases where the event was added through a XAML attribute, which means that the handler was added in generated code.
[...] For C++/CX, you can't remove handlers that were added as a XAML attribute because the C++/CX generated code doesn't save a token.



Is the below behaviour related to Uno-specific XAML-parsing then? The given caveat does not apply when running regularly on WinUi w/ C#/WinRT (as shown). Even Xamarin.Forms seems to have no issue with this (thus apparently also no restriction through Xamarin.Android/Xamarin.iOS).

> Which is why [...] the provided event handler is not removed because it is never registered directly, as instead it's invoked through a generated lambda.

**Reply by jeromelaban** [reply]
> Is the below behaviour related to Uno-specific XAML-parsing then? The given caveat does not apply when running regularly on WinUi w/ C#/WinRT (as shown). Even Xamarin.Forms seems to have no issue with this (thus apparently also no restriction through Xamarin.Android/Xamarin.iOS).
> 

That's correct, this is uno-specific. Uno needs this behavior to avoid memory leaks induced by attaching handlers.
